### ISSUE TITLE
Plotting BDDs without complemented edges

### ISSUE THREAD

**Issue by blacktopol** [tulip-control/dd#41]
I have created BDD for 'x xor y' function. And it looks very strange.
It looks like TRUE xor TRUE = TRUE
[x_xor_y.pdf](https://github.com/tulip-control/dd/files/2691255/x_xor_y.pdf)

Do I have an ability to switch BDD to "classic form" (without negated low edge)?

**Comment by johnyf** [comment]
This possibility for plotting could be implemented.
- CUDD implements complemented edges, [see [the function `Cudd_IsComplement` in the file `cudd/cudd.h`](https://github.com/johnyf/cudd/blob/fa6c459cf304d185dd7b573e4dbba84f77070bef/cudd/cudd.h#L337-L347) and the [Section "New BDD and ADD Variables" in the CUDD User's Manual](https://github.com/johnyf/cudd/blob/fa6c459cf304d185dd7b573e4dbba84f77070bef/doc/cudd.tex.in#L444-L446) by Fabio Somenzi],
- Sylvan implements complemented edges [see page 28 of [Sylvan: multi-core decision diagrams](https://ris.utwente.nl/ws/portalfiles/portal/6047499/thesis_T_van__Dijk.pdf) by Tom van Dijk]
- BuDDy does not implement complemented edges (negated pointers), [see Section "Some Implementation details" of the [page "BuDDy: A BDD package" in the BuDDy Manual](http://buddy.sourceforge.net/manual/main.html) and [the function `not_rec` in the file `src/bddop.c`](https://github.com/utwente-fmt/buddy/blob/d04651cd1ae294538bf140864360f0eda0f87f1c/src/bddop.c#L425-L458) ]

- `dd.bdd`, and so `dd.autoref`, implements complemented edges.

The function `dd.bdd.to_pydot` is implemented for the class `dd.bdd.BDD` as the argument `bdd`:

https://github.com/tulip-control/dd/blob/a77dc26d7d08d45504bcb59c39eb6107e11bec26/dd/bdd.py#L1792-L1804

This function is called by the method `dd.bdd.BDD.dump`:

https://github.com/tulip-control/dd/blob/a77dc26d7d08d45504bcb59c39eb6107e11bec26/dd/bdd.py#L1179-L1208

and so by the method `dd.autoref.BDD.dump`:

https://github.com/tulip-control/dd/blob/a77dc26d7d08d45504bcb59c39eb6107e11bec26/dd/autoref.py#L191-L196

and by the method `dd.cudd.BDD.dump` (which first copies the BDD nodes reachable from the given root to an instance of the class `dd.autoref.BDD.dump`):

https://github.com/tulip-control/dd/blob/a77dc26d7d08d45504bcb59c39eb6107e11bec26/dd/cudd.pyx#L1069-L1079

One possibility would be to adapt this function to plotting BDDs from BuDDy (after adding to `dd/buddy.pyx` access to the successors of a node).

Another possibility would be to traverse in Python a BDD with complemented edges (for example using `dd.cudd.BDD`), and construct a `networkx` graph for the same BDD represented without negated edges.

About the PDF `x_xor_y.pdf`, it is correct. The code to reproduce it is:

```python
from dd import autoref as _bdd

bdd = _bdd.BDD()
bdd.declare('x', 'y')
u = bdd.add_expr('x ^ y')
bdd.dump('x_xor_y.pdf', [u])
```

The reference `u` is a negated edge, and outside the BDD manager. That `u` is negated could be checked with

```python
>>> print(u.negated)
True
>>> print(u.node)
-4
```

So this node (`x XOR y`) is the negated function represented by the node with index 4 (the top node). So `TRUE XOR TRUE` is `TRUE` in this BDD, and `FALSE` from the negated edge `u`. In other words, the BDD plotted is of the function `NOT (x XOR y)`.

This issue suggests adding to the function `dd.bdd.to_pydot` plotting of the given root nodes, so that if external references are complemented edges, then this would be evident in the diagram.

**Comment by johnyf** [comment]
Relevant to #39, where plotting external references to BDDs arises too.

**Comment by johnyf** [comment]
Another possibility would be to install a version before https://github.com/tulip-control/dd/commit/0d4b92766d8301fb331163e8e30e51265bd94043, the commit which implemented complemented edges for the module `dd.bdd`. The function `dd.bdd.to_pydot` before that change was

https://github.com/tulip-control/dd/blob/5b62882267d08528bc4b4cd9b7b391999a041fe6/dd/bdd.py#L774-L820

**Comment by johnyf** [comment]
Plotting of external references to BDD nodes has been implemented in 5871b7b9af5242e8afa49c36f9425e329b86aa78.

**Comment by johnyf** [comment]
A function for plotting BDDs without complemented edges is the following. It was obtained by modifying the function `dd.bdd.to_pydot`.

```python
# Copyright 2014-2020 by California Institute of Technology
# All rights reserved. Licensed under BSD-3.
#
import pydot


def to_pydot(roots, bdd):
    """Convert `BDD` to pydot graph.

    Nodes are ordered by variable levels in support.
    Edges to low successors are dashed.

    Nodes not reachable from `roots`
    are ignored, unless `roots is None`.

    The roots are plotted as external references.

    @type roots: container of BDD nodes
    @type bdd: `BDD`
    """
    # all nodes ?
    if roots is None:
        nodes = bdd._succ
        roots = list()
    else:
        nodes = bdd.descendants(roots)
    # show only levels in aggregate support
    levels = {bdd._succ[abs(u)][0] for u in nodes}
    assert bdd._succ[1][0] in levels
    g = pydot.Dot('bdd', graph_type='digraph')
    skeleton = list()
    subgraphs = dict()
    # layer for external BDD references
    layers = [-1] + sorted(levels)
    # add nodes for BDD levels
    for i in layers:
        h = pydot.Subgraph('', rank='same')
        g.add_subgraph(h)
        subgraphs[i] = h
        # add phantom node
        u = 'L{i}'.format(i=i)
        skeleton.append(u)
        if i == -1:
            # layer for external BDD references
            label = 'ref'
        else:
            # BDD level
            label = str(i)
        nd = pydot.Node(name=u, label=label, shape='none')
        h.add_node(nd)
    # auxiliary edges for ranking
    for i, u in enumerate(skeleton[:-1]):
        v = skeleton[i + 1]
        e = pydot.Edge(str(u), str(v), style='invis')
        g.add_edge(e)
    # add nodes
    idx2var = {k: v for v, k in bdd.vars.items()}
    # BDD nodes
    def f(x):
        return str(abs(x))
    for u in nodes:
        i, v, w = bdd._succ[abs(u)]
        # terminal ?
        if v is None:
            var = str(bool(abs(u)))
        else:
            var = idx2var[i]
        su = f(u)
        neg_su = '-' + su
        label = '{var}-{u}'.format(var=var, u=su)
        nd = pydot.Node(name=su, label=label)
        neg_nd = pydot.Node(name=neg_su, label='-' + label)
        # add node to subgraph for level i
        h = subgraphs[i]
        h.add_node(nd)
        h.add_node(neg_nd)
        # add edges
        if v is None:
            continue
        sv = f(v)
        sw = f(w)
        neg_sv = '-' + sv
        neg_sw = '-' + sw
        if v < 0:
            sv, neg_sv = neg_sv, sv
        # edges from positive `u`
        e = pydot.Edge(su, sv, style='dashed')
        g.add_edge(e)
        e = pydot.Edge(su, sw, style='solid')
        g.add_edge(e)
        # edges from negated `u`
        e = pydot.Edge(neg_su, neg_sv, style='dashed')
        g.add_edge(e)
        e = pydot.Edge(neg_su, neg_sw, style='solid')
        g.add_edge(e)
    # external references to BDD nodes
    for u in roots:
        i, _, _ = bdd._succ[abs(u)]
        su = 'ref' + str(u)
        label = '@' + str(u)
        nd = pydot.Node(name=su, label=label)
        # add node to subgraph for level -1
        h = subgraphs[-1]
        h.add_node(nd)
        # add edge from external reference to BDD node
        assert u is not None
        sv = str(abs(u))
        if u < 0:
            sv = '-' + sv
        e = pydot.Edge(su, sv, style='dashed')
        g.add_edge(e)
    return g
```

The XOR example (described also [above](https://github.com/tulip-control/dd/issues/41#issuecomment-589183697)) can be plotted as follows, using the function `to_pydot` from above.

```python
from dd import autoref as _bdd

bdd = _bdd.BDD()
bdd.declare('x', 'y')
u = bdd.add_expr('x ^ y')

roots = [u.node]
g = to_pydot(roots, bdd._bdd)
g.write_pdf('x_xor_y.pdf')
```

The result is the following diagram

![x_xor_y](https://user-images.githubusercontent.com/3135273/75093542-41254c80-558b-11ea-9c92-507d4a2ce548.png)
